db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/server/DiagramDbServiceHandler.java

```diff
@@ -98,6 +131,63 @@ public TFolder getFolderTree(String username) throws TNotFound {
         } catch (NotFoundException e) {
             throw new TNotFound(username, "FolderTree for specified user not found.");
         }
-        return folder.toTFolder();
+        return folder.toTFolder(username);
+    }
+
+    @Override
+    public void shareFolderTo(String username, TFolder tFolderToShare) throws TException {
+        Folder rootFolder = mapper.convertTFolder(getFolderTree(username));
+
+        Folder folderToShare = mapper.convertTFolder(tFolderToShare);
+        folderToShare = addOwner(username, folderToShare);
+        updateFolder(folderToShare.toTFolder(username));
+
+        if (rootFolder == null) {
+            return;
+        }
+
+        Folder sharedFolder = getSharedFolder(rootFolder);
+
+        if (sharedFolder == null) {
+
+            sharedFolder = new Folder("Shared", username);
+            long idShared = saveFolder(sharedFolder.toTFolder(username));
+            sharedFolder.setId(idShared);
+
+            sharedFolder.getChildrenFolders().add(folderToShare);
+            folderToShare.getParentFolders().add(sharedFolder);
+
+            updateFolder(sharedFolder.toTFolder(username));
+
+            rootFolder.getChildrenFolders().add(sharedFolder);
+            sharedFolder.getParentFolders().add(rootFolder);
+            updateFolder(rootFolder.toTFolder(username));
+        }
+        else {
+            if (sharedFolder.getChildrenFolders().contains(folderToShare)) {
+                logger.error("addUserToOwners method was called with folder which already shared to user.");
+                return;
+            }
+            sharedFolder.getChildrenFolders().add(folderToShare);
+            updateFolder(sharedFolder.toTFolder(username));
+        }
+    }
+
+    private Folder addOwner(String username, Folder folder) {
+        Set<String> owners = folder.getOwners();
+        owners.add(username);
+        folder.setOwners(owners);
+        return folder;
```

Comment:
No, it's a good practice to mark arguments that are not supposed to be changed as `final`. If the argument is not final, it could be changed by design.